Issue 5 - Move HIE Tests and convert to Tasty

jeffwindsor opened the pull request:
Tests are converted to Tasty format.  The tests compile and run from console, but most fail / timeout.
I want to check in to see if this is going the right direction, and maybe get some help wiring up the CI / Runner / Config.

jeffwindsor commented:
helps resolve #5

alanz commented:
I'm a bit embarrased to say this, but we decided to stick with hspec, because it has better rerun options.

Bodigrim commented:
What's missing in `tasty-rerun`?

alanz commented:
@Bodigrim possibly just my inability to look up documentation.

But I do know that if an hspec test fails, it lists a way to rerun just that test, with the appropriate command line invocation, as part of the test result. I find this invaluable.

alanz commented:
And, full disclosure, I have a personal bias, as all my other OSS projects use hspec.

Bodigrim commented:
`tasty-rerun` provides `--rerun` and `--rerun-all-on-success` options, pretty much similar to `hspec`. 

Full disclosure: I also have a personal bias, being a maintainer of `tasty-rerun` :) I do not mean to interfere with your project's decisions, but I'm genuinely interested what can be improved. If something is missing from `tasty-rerun`, I'll be happy to fix it.

jeffwindsor commented:
closing to move to hspec

fendor commented:
More discussion needed. As I look at this PR, I dont have anything to complain about on the first sight! 

@jeffwindsor Please dont undo your work, yet.

alanz commented:
@Bodigrim glad to have you take an interest. The main thing I care about is if a test fails, I get something like

```
  To rerun use: --match "/Plugin/composes code actions/provides 3.8 code actions/"
```
And then I can do 

```bash
stack --stack-yaml=stack.yaml test haskell-language-server:func-test --test-arguments "--match \"/Plugin/composes code actions/provides 3.8 code actions/\""
```
To run just that test

alanz commented:
One other thing that would be useful, is junit.xml output. as per https://github.com/haskell/haskell-ide-engine/blob/master/test/utils/TestUtils.hs#L306

Bodigrim commented:
> One other thing that would be useful, is junit.xml output

http://hackage.haskell.org/package/tasty-ant-xml works well for me.

alanz commented:
To summarize the discussion on IRC, we will continue with tasty.  For the avoidance of future doubt. Sorry to have caused so much trauma for all.

fendor commented on haskell-language-server.cabal:
```diff
@@ -205,69 +205,59 @@ executable haskell-language-server-wrapper
   import: agpl
   hs-source-dirs:      test/utils
-  exposed-modules:     TestUtils
+  exposed-modules:     Test.HIE.Util
```
This should be `Test.Hls.Util`? Or `Test.HLS.Util`, but personally, I like the former more

jeffwindsor replied:
added to next push

fendor commented on haskell-language-server.cabal:
```diff
@@ -205,69 +205,59 @@ executable haskell-language-server-wrapper
   hs-source-dirs:      test/utils
-  exposed-modules:     TestUtils
+  exposed-modules:     Test.HIE.Util
+                     , Test.Tasty.Expectation
```
This module has been deelted.

jeffwindsor replied:
done

fendor commented on haskell-language-server.cabal:
```diff
@@ -205,69 +205,59 @@ executable haskell-language-server-wrapper
+  type:                 exitcode-stdio-1.0
+  default-language:     Haskell2010
+  build-tool-depends:   haskell-language-server:haskell-language-server
+                      , cabal-helper:cabal-helper-main
```
The whole line is in my point of view unneeded. 
cc @alanz or did you add it for a reason?

alanz replied:
Well, we are running functional tests that actually run the server. If the test does not depend on the server, we are likely to end up running a functional tests against a stale build of the server.

I did not add it, but I think that is why.

fendor replied:
But this `cabal-helper:cabal-helper-main` should be unnecassary, right?

fendor replied:
Looks to me light an artefact from pre cabal-helper-1.0

alanz replied:
If we are not using cabal-helper for running the server, then it can be removed

jeffwindsor replied:
done

fendor commented on test/utils/Test/HIE/Util.hs:
```diff
@@ -1,24 +1,27 @@
 {-# LANGUAGE CPP, OverloadedStrings, NamedFieldPuns #-}
-module TestUtils
+module Test.HIE.Util
```
Out of interest, why is this renamed?

jeffwindsor replied:
was looking for more consistent namespacing with the other testing tools, but really don't have a strong opinion.  easy for me to put but if you prefer the other way

fendor replied:
No, was just asking, dont have a preference either. Keep it :)

jeffwindsor replied:
done Test.Hls.Util

jeffwindsor commented:
@fendor: 

- broken tests are marked and "pass" (https://github.com/nomeata/tasty-expected-failure/blob/master/Test/Tasty/ExpectedFailure.hs)
- added xml output (https://github.com/ocharles/tasty-ant-xml/blob/master/Test/Tasty/Runners/AntXML.hs)
- added rerun (https://github.com/ocharles/tasty-rerun/blob/master/src/Test/Tasty/Ingredients/Rerun.hs)

as of this commit I have not test the re-run or xml output successfully, but I am sure it is just a matter of getting the cli parameters to get through with the cradle

fendor commented:
Awesome, I will review it in the next days in depth!

Bodigrim commented:
> @Bodigrim glad to have you take an interest. The main thing I care about is if a test fails, I get something like
> 
> ```
>   To rerun use: --match "/Plugin/composes code actions/provides 3.8 code actions/"
> ```

Your wish is [granted](https://hackage.haskell.org/package/tasty-1.4.2/changelog) :)
